### ISSUE TITLE
Can't replace platform_package framework-arduinoespressif32, "waiting for download"

### ISSUE THREAD

**Issue by ma-lalonde** [esphome/issues#3542]
### The problem

I need to get Ethernet working with a single-core ESP32-MINI-1. To this effect, I have compiled **and successfully tested** a single-core version of arduino-esp32. I am now trying to get it working with ESPHome, but after successfully compiling the .yaml file and uploading it to the MCU, I get stuck at "waiting for download" after reboot.

I am looking for pointers on how to solve this, if it has been done before, if there is a way to get more feedback from the terminal.

I have also read https://github.com/esphome/feature-requests/issues/1683 and https://github.com/esphome/issues/issues/1560, and am aware of the [platformio documentation](https://esphome.io/components/esphome.html#esphome-platformio-options).

I have successfully had my board running (with only WiFi) using the following:
``` yaml
esp32:
  board: esp32dev
  framework:
    type: esp-idf
    sdkconfig_options:
      CONFIG_FREERTOS_UNICORE: y
```
Unfortunately, as mentioned by the issue above, Ethernet is only implemented through Arduino right now. A quick "getting started" tutorial on how I could implement Ethernet for ESP-IDF would also be an acceptable solution.

### Which version of ESPHome has the issue?

2022.8.1

### What type of installation are you using?

pip

### Which version of Home Assistant has the issue?

2022.8.7

### What platform are you using?

ESP32

### Board

Espoir

### Component causing the issue

Arduino

### Example YAML snippet

```yaml
esphome:
  name: espoir001
  platformio_options:
    platform: espressif32
    platform_packages:
      - framework-arduinoespressif32 @ https://github.com/Connaxio/arduino-esp32/archive/refs/tags/2.0.4-1.zip

esp32:
  board: esp32dev
  framework: 
    type: arduino
```

### Anything in the logs that might be useful for us?

```txt
esphome run espoir001.yaml
INFO Reading configuration espoir001.yaml...
INFO Generating C++ source...
INFO Compiling app...
********************************************************************************
Obsolete PIO Core v6.0.2 is used (previous was 6.1.4)
Please remove multiple PIO Cores from a system:
https://docs.platformio.org/en/latest/core/installation/troubleshooting.html
********************************************************************************
Processing espoir001 (board: esp32dev; framework: arduino; platform: espressif32)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - framework-arduinoespressif32 @ 2.0.4 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch3
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
Dependency Graph
|-- WiFi @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- Update @ 2.0.0
|-- noise-c @ 0.1.4
|   |-- libsodium @ 1.10018.1
RAM:   [=         ]  11.5% (used 37528 bytes from 327680 bytes)
Flash: [=====     ]  45.7% (used 837929 bytes from 1835008 bytes)
========================= [SUCCESS] Took 2.38 seconds =========================
INFO Successfully compiled program.
Found multiple options, please choose one:
  [1] /dev/ttyUSB0 (CP2102N USB to UART Bridge Controller)
  [2] Over The Air (192.168.1.119)
(number): 1
esptool.py v3.3.1
Serial port /dev/ttyUSB0
Connecting....
Chip is ESP32-U4WDH (revision 3)
Features: WiFi, BT, Single Core, Embedded Flash, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: 1c:9d:c2:04:56:ac
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
Configuring flash size...
Auto-detected Flash size: 4MB
Flash will be erased from 0x00010000 to 0x000ddfff...
Flash will be erased from 0x00001000 to 0x00005fff...
Flash will be erased from 0x00008000 to 0x00008fff...
Flash will be erased from 0x0000e000 to 0x0000ffff...
Compressed 843696 bytes to 561232...
Wrote 843696 bytes (561232 compressed) at 0x00010000 in 13.7 seconds (effective 494.3 kbit/s)...
Hash of data verified.
Flash params set to 0x0020
Compressed 17216 bytes to 12006...
Wrote 17216 bytes (12006 compressed) at 0x00001000 in 0.5 seconds (effective 254.3 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 144...
Wrote 3072 bytes (144 compressed) at 0x00008000 in 0.1 seconds (effective 441.1 kbit/s)...
Hash of data verified.
Compressed 8192 bytes to 47...
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.1 seconds (effective 602.5 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
INFO Successfully uploaded program.
INFO Starting log output from /dev/ttyUSB0 with baud rate 115200
[12:07:38]1150 mmu set 00010000, pos 00010000
[12:07:38]1150 mmu set 00020000, pos 00020000
[12:07:38]1150 mmu set 00030000, pos 00030000
[12:07:39]1150 mmu set 00040000, pos 00040000
[12:07:39]1150 mmu set 00050000, pos 00050000
[12:07:39]1150 mmu set 00060000, pos 00060000
[12:07:39]1150 mmu set 00070000, pos 00070000
[12:07:39]1150 mmu set 00080000, pos 00080000
[12:07:39]1150 mmu set 00090000, pos 00090000
[12:07:39]1150 mmu set 000a0000, pos 000a0000
[12:07:39]1150 mmu set 000b0000, pos 000b0000
[12:07:39]1150 mmu set 000c0000, pos 000c0000
[12:07:39]1150 mmu set 000d0000, pos 000d0000
[12:07:39]1150 mmu set 000e0000, pos 000e0000
[12:07:39]1150 mmu set 000f0000, pos 000f0000
[12:07:39]1150 mmu set 00100000, pos 00100000\xfaets Jul 29 2019 12:21:46
[12:07:39]
[12:07:39]rst:0x10 (RTCWDT_RTC_RESET),boot:0x3 (DOWNLOAD_BOOT(UART0/UART1/SDIO_REI_REO_V2))
[12:07:39]waiting for download
```

### Additional information

I tried to solve the `Obsolete PIO Core v6.0.2 is used` message following the instructions, but `pip install esphome` keeps reinstalling the same version.

I have also tried specifying the `espressif32` and `Arduino` versions in the .yaml file, with no success.

**Comment by ma-lalonde** [comment]
I got a little more information from the terminal using [Tasmota's SOLO1 build](https://github.com/Jason2866/esp32-arduino-lib-builder/releases/tag/922), but still nothing functional. I don't know how to interpret the additional details.

```
esphome run espoir001.yaml
INFO Reading configuration espoir001.yaml...
INFO Generating C++ source...
INFO Compiling app...
********************************************************************************
Obsolete PIO Core v6.0.2 is used (previous was 6.1.4)
Please remove multiple PIO Cores from a system:
https://docs.platformio.org/en/latest/core/installation/troubleshooting.html
********************************************************************************
Processing espoir001 (board: esp32dev; framework: arduino; platform: espressif32)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - framework-arduinoespressif32 @ 2.0.4 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch3
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
Dependency Graph
|-- WiFi @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- Update @ 2.0.0
|-- noise-c @ 0.1.4
|   |-- libsodium @ 1.10018.1
RAM:   [=         ]  10.7% (used 35000 bytes from 327680 bytes)
Flash: [====      ]  38.5% (used 706345 bytes from 1835008 bytes)
========================= [SUCCESS] Took 2.19 seconds =========================
INFO Successfully compiled program.
Found multiple options, please choose one:
  [1] /dev/ttyUSB0 (CP2102N USB to UART Bridge Controller)
  [2] Over The Air (192.168.1.119)
(number): 1
esptool.py v3.3.1
Serial port /dev/ttyUSB0
Connecting....
Chip is ESP32-U4WDH (revision 3)
Features: WiFi, BT, Single Core, Embedded Flash, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: 1c:9d:c2:04:56:ac
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
Configuring flash size...
Auto-detected Flash size: 4MB
Flash will be erased from 0x00010000 to 0x000bdfff...
Flash will be erased from 0x00001000 to 0x00004fff...
Flash will be erased from 0x00008000 to 0x00008fff...
Flash will be erased from 0x0000e000 to 0x0000ffff...
Compressed 712112 bytes to 506306...
Wrote 712112 bytes (506306 compressed) at 0x00010000 in 11.9 seconds (effective 477.8 kbit/s)...
Hash of data verified.
Flash params set to 0x0020
Compressed 15584 bytes to 10960...
Wrote 15584 bytes (10960 compressed) at 0x00001000 in 0.5 seconds (effective 265.4 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 144...
Wrote 3072 bytes (144 compressed) at 0x00008000 in 0.1 seconds (effective 443.6 kbit/s)...
Hash of data verified.
Compressed 8192 bytes to 47...
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.1 seconds (effective 646.4 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
INFO Successfully uploaded program.
INFO Starting log output from /dev/ttyUSB0 with baud rate 115200
[14:16:51]1150 mmu set 00010000, pos 00010000
[14:16:51]1150 mmu set 00020000, pos 00020000
[14:16:51]1150 mmu set 00030000, pos 00030000
[14:16:51]ets Jul 29 2019 12:21:46
[14:16:51]
[14:16:51]rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
[14:16:51]configsip: 188777542, SPIWP:0xee
[14:16:51]clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
[14:16:51]mode:QIO, clock div:2
[14:16:51]load:0x3fff0030,len:184
[14:16:51]load:0xfffcddfd,len:-3355443
[14:16:51]1150 mmu set 00010000, pos 00010000
[14:16:51]1150 mmu set 00020000, pos 00020000
[14:16:51]1150 mmu set 00030000, pos 00030000
[14:16:52]ets Jul 29 2019 12:21:46
[14:16:52]
[14:16:52]rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
[14:16:52]configsip: 188777542, SPIWP:0xee
[14:16:52]clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
[14:16:52]mode:QIO, clock div:2
[14:16:52]load:0x3fff0030,len:184
[14:16:52]load:0xfffcddfd,len:-3355443
[14:16:52]1150 mmu set 00010000, pos 00010000
[14:16:52]1150 mmu set 00020000, pos 00020000
[14:16:52]1150 mmu set 00030000, pos 00030000
[14:16:52]ets Jul 29 2019 12:21:46
[14:16:52]
[14:16:52]rst:0x10 (RTCWDT_RTC_RESET),boot:0x3 (DOWNLOAD_BOOT(UART0/UART1/SDIO_REI_REO_V2))
[14:16:52]waiting for download
```

**Comment by ma-lalonde** [comment]
This is probably related to the new merged bootloader format in Arduino 2.0.4. Closing
